allennlp/models/graph_parser.py

```diff
@@ -0,0 +1,341 @@
+        """
+        Parameters
+        ----------
+        tokens : Dict[str, torch.LongTensor], required
+            The output of ``TextField.as_array()``.
+        pos_tags : ``torch.LongTensor``, optional, (default = None).
+            The output of a ``SequenceLabelField`` containing POS tags.
+        arc_tags : torch.LongTensor, optional (default = None)
+            A torch tensor representing the sequence of integer indices denoting the parent of every
+            word in the dependency parse. Has shape ``(batch_size, sequence_length, sequence_length)``.
+
+        Returns
+        -------
+        An output dictionary.
+        """
+        embedded_text_input = self.text_field_embedder(tokens)
+        if pos_tags is not None and self._pos_tag_embedding is not None:
+            embedded_pos_tags = self._pos_tag_embedding(pos_tags)
+            embedded_text_input = torch.cat([embedded_text_input, embedded_pos_tags], -1)
+        elif self._pos_tag_embedding is not None:
+            raise ConfigurationError("Model uses a POS embedding, but no POS tags were passed.")
+
+        mask = get_text_field_mask(tokens)
+        embedded_text_input = self._input_dropout(embedded_text_input)
+        encoded_text = self.encoder(embedded_text_input, mask)
+
+        float_mask = mask.float()
+        encoded_text = self._dropout(encoded_text)
+
+        # shape (batch_size, sequence_length, arc_representation_dim)
+        head_arc_representation = self._dropout(self.head_arc_feedforward(encoded_text))
+        child_arc_representation = self._dropout(self.child_arc_feedforward(encoded_text))
+
+        # shape (batch_size, sequence_length, tag_representation_dim)
+        head_tag_representation = self._dropout(self.head_tag_feedforward(encoded_text))
+        child_tag_representation = self._dropout(self.child_tag_feedforward(encoded_text))
+        # shape (batch_size, sequence_length, sequence_length)
+        attended_arcs = self.arc_attention(head_arc_representation,
+                                           child_arc_representation)
+        # shape (batch_size, num_tags, sequence_length, sequence_length)
+        arc_tag_logits = self.tag_bilinear(head_tag_representation,
+                                           child_tag_representation)
+        # Switch to (batch_size, sequence_length, sequence_length, num_tags)
+        arc_tag_logits = arc_tag_logits.permute(0, 2, 3, 1).contiguous()
+
+        minus_inf = -1e8
+        minus_mask = (1 - float_mask) * minus_inf
+        attended_arcs = attended_arcs + minus_mask.unsqueeze(2) + minus_mask.unsqueeze(1)
+
+        arc_probs, arc_tag_probs = self._greedy_decode(attended_arcs,
+                                                       arc_tag_logits,
+                                                       mask)
+        # Make the arc tags not have negative values anywhere
+        # (by default, no edge is indicated with -1).
+        arc_indices = (arc_tags != -1).float()
```

Comment:
This block needs to go inside `if arg_tags is not None`, right?

Reply:
Yep, fixed thanks